middleware/camera/views.py

```
@@ -0,0 +1,269 @@
+            logger.error("An exception occurred while getting presets: %s", exc)
+            return Response(
+                {"message": exc.default_detail}, status=status.HTTP_400_BAD_REQUEST
+            )
+
+    @extend_schema(
+        request=CameraAssetPresetRequest,
+        responses={200: "string"},
+        description="Set a preset",
+    )
+    @action(detail=False, methods=["post"], url_name="presets")
+    def set_preset(self, request):
+        try:
+            cam_request = CameraAssetPresetRequest.model_validate(request.data)
+            cam = OnvifZeepCameraController(req=cam_request)
+            result = cam.set_preset(preset_name=cam_request.preset_name)
+            return Response(result, status=status.HTTP_200_OK)
+        except ValidationError as exc:
+            return Response(
+                {"message": exc.errors()}, status=status.HTTP_400_BAD_REQUEST
+            )
+        except InvalidCameraCredentialsException as exc:
+            logger.error("An exception occurred while getting presets: %s", exc)
+            return Response(
+                {"message": exc.default_detail}, status=status.HTTP_400_BAD_REQUEST
+            )
+
+    @extend_schema(
+        request=CameraAssetPresetRequest,
+        responses={200: "string"},
+        description="go to a particular preset",
+    )
+    @action(detail=False, methods=["post"], url_path="gotoPreset")
+    def go_to_preset(self, request):
+        try:
+            cam_request = CameraAssetPresetRequest.model_validate(request.data)
+        except ValidationError as e:
+            return Response(e.errors(), status=status.HTTP_400_BAD_REQUEST)
+
+        if is_camera_locked(cam_request.hostname):
+            logger.info("Camera with host: %s is locked.", cam_request.hostname)
+            return Response(
+                {
+                    "message": "Camera is Locked!",
+                },
+                status=status.HTTP_423_LOCKED,
+            )
```

extract this to a helper function in the viewset 

```
def check_camera_state(self, id, raise=False):
	state = is_camera_locked(id)
	logger.debug(...state...)
	if state and raise:
		raise DRFValidationError(code=423, {...})
	return state
```

then just call this function wherever you need to just raise an error